control/discovery.py

```diff
@@ -0,0 +1,1127 @@
+        nvme_tcp_data_pdu.cmd_id = cmd_id
+        nvme_tcp_data_pdu.data_length = nvme_data_len
+
+        # reply based on the received get log page request packet(length)
+        if nvme_data_len == 16:
+            # nvme cli version: 1.x
+            nvme_get_log_page_reply = NVMeGetLogPage()
+            nvme_get_log_page_reply.genctr = self_conn.gen_cnt
+            nvme_get_log_page_reply.numrec = len(listeners)
+
+            reply = bytes(pdu_reply) + bytes(nvme_tcp_data_pdu) + \
+                    bytes(nvme_get_log_page_reply)[:16]
+        elif nvme_data_len == 1024 and nvme_logpage_offset == 0:
+            # nvme cli version: 2.x
+            nvme_get_log_page_reply = NVMeGetLogPage()
+            nvme_get_log_page_reply.genctr = self_conn.gen_cnt
+            nvme_get_log_page_reply.numrec = len(listeners)
+
+            reply = bytes(pdu_reply) + bytes(nvme_tcp_data_pdu) + \
+                    bytes(nvme_get_log_page_reply)
+        elif nvme_data_len % 1024 == 0:
+            # reply log pages
+            reply = bytes(pdu_reply) + bytes(nvme_tcp_data_pdu) + \
+                self_conn.log_page[nvme_logpage_offset:nvme_logpage_offset+nvme_data_len]
+            self_conn.unsent_log_page_len -= nvme_data_len
+            if self_conn.unsent_log_page_len == 0:
+                self_conn.log_page = b''
+                self_conn.property_set = False
+                self_conn.allow_listeners = []
+        else:
+            self.logger.error("lenghth error. It need to be 16 or n*1024")
+            return -1
+        try:
+            conn.sendall(reply)
+        except BrokenPipeError:
+            self.logger.error("client disconnected unexpectedly.")
+            return -1
+        self.logger.debug("reply get log page request.")
+        return 0
+
+    def reply_keep_alive(self, conn, data, cmd_id):
+        """Reply keep alive request."""
+
+        self.logger.debug("handle keep alive request.")
+        self_conn = self.conn_vals[conn.fileno()]
+        nvme_sgl = struct.unpack_from('<16B', data, 32)
+        nvme_sgl_desc_type = nvme_sgl[15] & 0xF0
+        nvme_sgl_desc_sub_type = nvme_sgl[15] & 0x0F
+        nvme_keep_alive_dword10 = struct.unpack_from('<I', data, 48)[0]
+        nvme_keep_alive_dword11 = struct.unpack_from('<I', data, 52)[0]
+        nvme_keep_alive_dword12 = struct.unpack_from('<I', data, 56)[0]
+        nvme_keep_alive_dword13 = struct.unpack_from('<I', data, 60)[0]
+        nvme_keep_alive_dword14 = struct.unpack_from('<I', data, 64)[0]
+        nvme_keep_alive_dword15 = struct.unpack_from('<I', data, 68)[0]
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = b'\x18'
+        pdu_reply.packet_length = b'\x18\x00\x00\x00'
+
+         # Cqe for keep alive
+        keep_alive_reply = CqeNVMe()
+        keep_alive_reply.sq_head_ptr = struct.pack('<H', self_conn.sq_head_ptr)
+        keep_alive_reply.cmd_id = struct.pack('<H', cmd_id)
```

Comment:
By using [`ctypes.BigEndianStructure`](https://docs.python.org/3/library/ctypes.html#ctypes.BigEndianStructure), you probably don't need `pack`/`unpack`:
```suggestion
        keep_alive_reply = CqeNVMe(
            sq_head_ptr=self_conn.sq_head_ptr,
            cmd_id=cmd_id)
```

Reply:
Yes, I have already used `ctypes`, so there is no need to use `pack`. This is an omission. Just assign a value. In addition, the NVME protocol is `LittleEndianStructure` so in order to be more accuracy, I have changed the `Structure` to `LittleEndianStructure`.